[V1][Spec Decode] Handle draft tokens beyond max_model_len

WoosukKwon opened the pull request:
Implements 4. Handle the edge cases like when the draft model generates beyond max_pos_embeddings in #15901

WoosukKwon committed:
[Spec Decode] Do not generate draft tokens beyond max_model_len

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

comaniac reviewed (approved):
Overall LGTM. Approve first to unblock the PR. Meanwhile it would be good to have a unit test for it. Also do we know the overhead of introduced ops (e.g., torch.where)?

WoosukKwon committed:
Fix ngram

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Update comments

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

ekagra-ranjan commented on vllm/v1/spec_decode/eagle.py:
```diff
@@ -20,7 +22,10 @@ def __init__(
+        self.max_model_len = vllm_config.model_config.max_model_len
         self.block_size = vllm_config.cache_config.block_size
+
+        # Optimization: pre-compute and cache the arange tensor.
```
maybe we could also pass in one used here: https://github.com/vllm-project/vllm/blob/027b204ff15e803190775e04d39973606a3a7021/vllm/v1/worker/gpu_model_runner.py#L240

ekagra-ranjan commented on vllm/v1/spec_decode/eagle.py:
```diff
@@ -103,22 +108,52 @@ def propose(
+            # Mask out the slot mappings that exceed the max model length.
+            # Otherwise, the KV cache will be inadvertently updated with the
+            # padding tokens.
+            attn_metadata.slot_mapping = torch.where(
+                exceeds_max_model_len,
+                PADDING_SLOT_ID,
+                attn_metadata.slot_mapping,
+            )
```
My understanding is that torch.where will allocate an intermedia tensor and then assign it. Is it possible to use `attn_metadata.slot_mapping[exceeds_max_model_len] = PADDING_SLOT_ID` to that it's an in-place operation?

WoosukKwon replied:
@ekagra-ranjan Thanks for the suggestion. I changed it to `masked_fill_`, which is also an in-place operation.
Overall, I think the performance impact will be small since the tensors here are small (shape of `[batch_size]`).

ekagra-ranjan commented on vllm/v1/spec_decode/eagle.py:
```diff
@@ -103,22 +108,52 @@ def propose(
+            # generates tokens beyond the max model length. Since it is complex
+            # to remove such requests from the batch, we keep them in the batch
+            # but adjust the position ids and slot mappings to avoid the
+            # out-of-range access during the model execution. The draft tokens
+            # generated with this adjustment should be ignored.
```
Do we need to add the logic to ignoring such draft tokens [here](https://github.com/vllm-project/vllm/pull/16087/files#diff-80ee7e2a62f9dcfbb8a312dc4e3948557e97ef187290daebbcae1e28596bda29R1200)?

WoosukKwon replied:
Good question. The scheduler handles it https://github.com/vllm-project/vllm/blob/af7462b6d1ebc10ca83da57f2804c9e49a40e83c/vllm/v1/core/sched/scheduler.py#L188-L193

WoosukKwon committed:
Merge branch 'main' into eagle-max-len

WoosukKwon committed:
Add test

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Add comment

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon commented:
@comaniac Good point. Added a test. Also, I replaced `torch.where` with `masked_fill_` wherever possible, for better performance. Overall, I think the overhead will be very small because the masked tensors are all small (i.e., shape of `[batch_size]`).

WoosukKwon committed:
Fix test

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Fix test

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Merge branch 'main' into eagle-max-len

WoosukKwon committed:
Merge branch 'main' into eagle-max-len

WoosukKwon committed:
minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
fix ngram test

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>
